automerge-c/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+/// doc must be a pointer to a valid AMdoc
+/// key and value must be valid c strings
+#[no_mangle]
+pub unsafe extern "C" fn AMconfig(
+    doc: *mut AMdoc,
+    key: *const c_char,
+    value: *const c_char,
+) -> *mut AMresult {
+    let doc = to_doc!(doc);
+    let key = to_str!(key);
+    match key.as_str() {
+        "actor" => {
+            let actor = to_str!(value);
+            if let Ok(actor) = actor.try_into() {
+                doc.set_actor(actor);
+                AMresult::Ok.into()
+            } else {
+                AMresult::err(&format!("Invalid actor '{}'", to_str!(value))).into()
+            }
+        }
+        k => AMresult::err(&format!("Invalid config key '{}'", k)).into(),
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Get an `AMdoc` struct's actor ID value as a hexadecimal string.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \return A pointer to an `AMresult` struct containing a string value.
+/// \pre \p doc must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+#[no_mangle]
+pub unsafe extern "C" fn AMgetActor(_doc: *mut AMdoc) -> *mut AMresult {
+    unimplemented!()
+}
+
+/// \memberof AMresult
+/// \brief Get the status code of an `AMresult` struct.
+///
+/// \param[in] result A pointer to an `AMresult` struct or `NULL`.
+/// \return An `AmStatus` enum tag.
+///
+/// # Safety
+/// result must be a pointer to a valid AMresult
+#[no_mangle]
+pub unsafe extern "C" fn AMresultStatus(result: *mut AMresult) -> AmStatus {
+    match result.as_mut() {
+        Some(AMresult::Ok) => AmStatus::CommandOk,
+        Some(AMresult::Error(_)) => AmStatus::Error,
+        Some(AMresult::ObjId(_)) => AmStatus::ObjOk,
+        Some(AMresult::Values(_)) => AmStatus::ValuesOk,
+        Some(AMresult::Changes(_)) => AmStatus::ChangesOk,
+        None => AmStatus::InvalidResult,
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Set a map object's value.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] obj A pointer to an `AMobj` struct.
+/// \param[in] key A map object's key string.
+/// \param[in] data_type An `AmDataType` enum tag matching the actual type that
+///            \p value points to.
+/// \param[in] value A pointer to the value at \p key or `NULL`.
+/// \return A pointer to an `AMresult` struct containing no value.
+/// \pre \p doc must be a valid address.
+/// \pre \p obj must be a valid address.
+/// \pre \p key must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+/// obj must be a pointer to a valid AMobj or NULL
+/// key must be a c string of the map key to be used
+/// value must be a pointer to data of the type specified in data_type
+#[no_mangle]
+pub unsafe extern "C" fn AMmapSet(
```

Comment:
Here we replace any unicode characters in the value (if it's a string) with U+FFFD, we should probably document this.